ESPHome 2025.8.0 fails to compile when using `on_scan_end` trigger of  `esp32_ble_tracker`

### The problem

My BLE proxies which utilise the on_scan_end trigger of esp32_ble_tracker fail to compile on 2025.8.0 but are fine on previous versions.

example yaml

```
esp32_ble_tracker:
  scan_parameters:
    active: false
  on_scan_end:
    - then:
      - script.execute: restart_after_delay
```

Compile error:

```
...
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/bta/sys/bta_sys_conn.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/bta/sys/bta_sys_main.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/bta/sys/utl.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/btc/core/btc_ble_storage.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/btc/core/btc_config.c.o
src/main.cpp:28:20: error: 'BLEEndOfScanTrigger' in namespace 'esphome::esp32_ble_tracker' does not name a type
   28 | esp32_ble_tracker::BLEEndOfScanTrigger *esp32_ble_tracker_bleendofscantrigger_id;
      |                    ^~~~~~~~~~~~~~~~~~~
src/main.cpp: In function 'void setup()':
src/main.cpp:297:3: error: 'esp32_ble_tracker_bleendofscantrigger_id' was not declared in this scope; did you mean 'esp32_ble_tracker_esp32bletracker_id'?
  297 |   esp32_ble_tracker_bleendofscantrigger_id = new esp32_ble_tracker::BLEEndOfScanTrigger(esp32_ble_tracker_esp32bletracker_id);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |   esp32_ble_tracker_esp32bletracker_id
src/main.cpp:297:50: error: expected type-specifier
  297 |   esp32_ble_tracker_bleendofscantrigger_id = new esp32_ble_tracker::BLEEndOfScanTrigger(esp32_ble_tracker_esp32bletracker_id);
      |                                                  ^~~~~~~~~~~~~~~~~
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/btc/core/btc_dev.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/btc/core/btc_dm.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/btc/core/btc_main.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/btc/core/btc_profile_queue.c.o
...
```

### Which version of ESPHome has the issue?

2025.8.0

### What type of installation are you using?

Home Assistant Add-on

### What platform are you using?

ESP32

### Component causing the issue

_No response_

### YAML Config

```yaml
esp32_ble_tracker:
  scan_parameters:
    active: false
  on_scan_end:
    - then:
      - script.execute: restart_after_delay
```

### Anything in the logs that might be useful for us?

```txt
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/bta/sys/bta_sys_conn.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/bta/sys/bta_sys_main.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/bta/sys/utl.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/btc/core/btc_ble_storage.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/btc/core/btc_config.c.o
src/main.cpp:28:20: error: 'BLEEndOfScanTrigger' in namespace 'esphome::esp32_ble_tracker' does not name a type
   28 | esp32_ble_tracker::BLEEndOfScanTrigger *esp32_ble_tracker_bleendofscantrigger_id;
      |                    ^~~~~~~~~~~~~~~~~~~
src/main.cpp: In function 'void setup()':
src/main.cpp:297:3: error: 'esp32_ble_tracker_bleendofscantrigger_id' was not declared in this scope; did you mean 'esp32_ble_tracker_esp32bletracker_id'?
  297 |   esp32_ble_tracker_bleendofscantrigger_id = new esp32_ble_tracker::BLEEndOfScanTrigger(esp32_ble_tracker_esp32bletracker_id);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |   esp32_ble_tracker_esp32bletracker_id
src/main.cpp:297:50: error: expected type-specifier
  297 |   esp32_ble_tracker_bleendofscantrigger_id = new esp32_ble_tracker::BLEEndOfScanTrigger(esp32_ble_tracker_esp32bletracker_id);
      |                                                  ^~~~~~~~~~~~~~~~~
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/btc/core/btc_dev.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/btc/core/btc_dm.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/btc/core/btc_main.c.o
Compiling .pioenvs/btproxy-atom-lite/bt/host/bluedroid/btc/core/btc_profile_queue.c.o
```

### Additional information

_No response_

--- a/esphome/components/esp32_ble_tracker/automation.h
+++ b/esphome/components/esp32_ble_tracker/automation.h
@@ -80,14 +80,17 @@ class BLEManufacturerDataAdvertiseTrigger : public Trigger<const adv_data_t &>, 
   ESPBTUUID uuid_;
 };
 
+#endif  // USE_ESP32_BLE_DEVICE
+
 class BLEEndOfScanTrigger : public Trigger<>, public ESPBTDeviceListener {
  public:
   explicit BLEEndOfScanTrigger(ESP32BLETracker *parent) { parent->register_listener(this); }
 
+#ifdef USE_ESP32_BLE_DEVICE
   bool parse_device(const ESPBTDevice &device) override { return false; }
+#endif
   void on_scan_end() override { this->trigger(); }
 };
-#endif  // USE_ESP32_BLE_DEVICE
 
 template<typename... Ts> class ESP32BLEStartScanAction : public Action<Ts...> {
  public:

--- a/tests/components/esp32_ble_tracker/test-on-scan-end.esp32-idf.yaml
+++ b/tests/components/esp32_ble_tracker/test-on-scan-end.esp32-idf.yaml
@@ -1,0 +1,3 @@
+esp32_ble_tracker:
+  on_scan_end:
+    - logger.log: "Scan ended!"

